### ISSUE TITLE
UWP: Realm crashes in UWP background tasks

### ISSUE THREAD

**Issue by Mohsens22** [realm/realm-dotnet#1762]
## Goals

I was trying to use CRUD operations in background tasks.

## Expected Results

I expected to work, a it works on SQLite

## Actual Results

It makes the background task crash.

## Steps to Reproduce

Make a background task, do read operation or a write transaction, boom! crashes.

## Code Sample
Use this in a background task:
```csharp
instance.Write(() =>
            {
                instance.Add(entity);
            });
```

## Version of Realm and Tooling

- Realm Object Server Version: ?
- Flavor:
  - [Developer
- Server OS & Version: Cloud
- Client SDK Version: V3.0 
- Client OS & Version: Windows 10 RS4 release

**Comment by fealebenpae** [comment]
Can you post a stack trace and exception message, please?

**Comment by Mohsens22** [comment]
We normally can't get a proper stack trace from background tasks. But I will post results as soon as I get one.

**Comment by Mohsens22** [comment]
It seems to be a Threading issue. It only happens when app is open.

**Comment by Mohsens22** [comment]
Unhandled exception at 0x77B41AE7 in backgroundTaskHost.exe: Microsoft C++ exception: realm::MultipleSyncAgents at memory location 0x05E5DA08.

Unhandled exception at 0x00000000 in backgroundTaskHost.exe: 0xC0000005: Access violation executing location 0x00000000.

**Comment by Mohsens22** [comment]
![image](https://user-images.githubusercontent.com/22152065/44612847-43390e80-a821-11e8-96cc-6cf010f20288.png)
'backgroundTaskHost.exe' (CoreCLR: CoreCLR_UWP_Domain): Loaded 'C:\Users\mohse\source\repos\Denna2\Denna\Denna\bin\x86\Debug\AppX\System.Linq.Queryable.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
Exception thrown: 'System.Exception' in Realm.dll
An exception of type 'System.Exception' occurred in Realm.dll but was not handled in user code
Realm accessed from incorrect thread.

The thread 0x74e8 has exited with code 0 (0x0).
The thread 0x5cc0 has exited with code 0 (0x0).
The thread 0x7664 has exited with code 0 (0x0).
The thread 0x453c has exited with code 0 (0x0).
The thread 0x14b8 has exited with code 0 (0x0).
The thread 0x788c has exited with code 0 (0x0).
The thread 0x7a2c has exited with code 0 (0x0).
The thread 0x75ec has exited with code 0 (0x0).
The thread 0x3344 has exited with code 0 (0x0).
The thread 0x79f8 has exited with code 0 (0x0).
The thread 0x1ff0 has exited with code 0 (0x0).
The thread 0x4a98 has exited with code 0 (0x0).
The program '[11700] backgroundTaskHost.exe' has exited with code 1 (0x1).

**Comment by nirinchev** [comment]
Looks like you're trying to use the Realm instance across threads which is not supported. In your background task, you should open a new instance. Additionally, I imagine you should either avoid doing async work while the Realm instance is open or use a package like Nito.AsyncEx to ensure that continuations are scheduled on the same thread.

**Comment by Mohsens22** [comment]
How do I fix this?
Method is async, but I make a new instance per background task

**Comment by Mohsens22** [comment]
Update: 
This mostly happens when I want to access instance and app is open

**Comment by nirinchev** [comment]
The thing is, in background threads, there's no synchronization context, which means that continuations are scheduled on a random threadpool thread. So after every await, there's a significant chance the thread id changes and the Realm you had opened already can't be used. The easiest approach I can think of is using [Nito.AsyncEx.Context](https://www.nuget.org/packages/Nito.AsyncEx.Context/1.1.0) package and do something like what we do in [the tests](https://github.com/realm/realm-dotnet/blob/master/Tests/Tests.Shared/AsyncTests.cs#L81-L102):

```csharp
AsyncContext.Run(async () =>
{
    var realm = Realm.GetInstance(...);
    await DoSomething();
    realm.Write(() => ...);
});
```

**Comment by Mohsens22** [comment]
I'll try this and let you know about the results.

**Comment by nirinchev** [comment]
Did you try using AsyncContext?

**Comment by Mohsens22** [comment]
Yes. But I recommend you to try that too.

**Comment by nirinchev** [comment]
Not sure what you mean - we're using that approach in our tests and things work as expected. If you have a project or a code snippet that reproduces the problem, I can try and help you, but a screenshot of an exception is definitely not enough to do any serious debugging.

**Comment by Mohsens22** [comment]
I'm making a source code for you bro, so you can reproduce problem. Wait.

**Comment by Mohsens22** [comment]
https://github.com/Mohsens22/RelamTest
Here you are.
Run this project once.
Then close and run the project again
Go to 
Services.Background project
Quick action class and put a breakpoint here:
![image](https://user-images.githubusercontent.com/22152065/45160204-22e45900-b1fe-11e8-9341-0392338b81d8.png)

Then From the top switch the lifecycle to "QuickAction" so you can debug.
![image](https://user-images.githubusercontent.com/22152065/45160264-44454500-b1fe-11e8-86f2-8555bc2d0192.png)

Respond to this thread if you have any question or you need any help.

**Comment by nirinchev** [comment]
I'm unable to reproduce the error. I do notice, however that the background thread locks when attempting to Write data to the Realm, so definitely something strange is going on. Are you sure your sample reproduces the error that you have mentioned - `Realm accessed from incorrect thread`?

**Comment by Mohsens22** [comment]
The behavior is so strange.
When the background task fails, you can't start debugging till you restart windows !!!
![image](https://user-images.githubusercontent.com/22152065/45223391-5db4c280-b2cc-11e8-8c54-8932e75eb502.png)
You get this error
The fun fact is that when app is open background task fails
When the app isn't open the possibility of failure is less :| wtf :|
I'm working on that sample to get an error to report.

**Comment by Mohsens22** [comment]
update:
wtf VS triggers first breakpoint but doesn't hut second :| Surely it starts the operation but refuses finishing it :| And it triggers no message 
![image](https://user-images.githubusercontent.com/22152065/45223784-7a052f00-b2cd-11e8-9a00-7b4a0ef5b1c3.png)

**Comment by nirinchev** [comment]
This is consistent with my observations - it seems `realm.Write` blocks the thread and can't proceed after the first few invocations.

**Comment by Mohsens22** [comment]
What can we do?

**Comment by nirinchev** [comment]
At the moment - nothing. It's likely a native issue that will require more debugging - I'll put it in our backlog and we'll address it at a later stage.

**Comment by Mohsens22** [comment]
Please check read operation as well
I don't remember well, but I think in some cases read can create background crash

**Comment by nirinchev** [comment]
This should be resolved with 10.0.0-beta.5 and newer.
